### PR TITLE
fix(lint/useTemplate): preserve leading non-string addition

### PR DESCRIPTION
**Pull request by Conaclos**
## Summary

Fix [rome#4713](https://github.com/rome/tools/issues/4713).
Previously, the rule made the following suggestion:

```diff
- a + b + "px"
+ `${a}${b}px`
```

This breaks code where `a` and `b` are numbers.
This PR fixes the issue making the following suggestion:

```diff
- a + b + "px"
+ `${a + b}px`
 ```

I take the opportunity to fix other unnoticed bugs:

- we emitted DOLLAR_CURLY token instead of R_CURLY for the ending mark of a template element.
- we erroneously handled tagged templates. They are now handled as non-string value

The rule also now provides the following suggestion (this aligns our behavior to the eslint behavior):
 
```diff
- v + "\n"
+ `${v}\n`
```

I also improved the code by avoiding recursions and reduced the number of traversals from 5 to 2.
The code to check if the binary operation is convertible to a template is now complete. This avoids to use memory in the case where the operation is not convertible (the majority of the situations).

## Test Plan

New tests added.

### CODE REVIEW

**Comment by Conaclos** [comment]
I don't understand why I got [this error on CI](https://github.com/biomejs/biome/actions/runs/5990507122/job/16247894990?pr=70).

Generating the rule documentation on local doesn't trigger this error (I tried a second time removing first my `target` directory).

Note that I added [`DOLLAR_CURLY` to the token kind -> string mapping](https://github.com/biomejs/biome/pull/70/files#diff-0065bccc1147cd025d6848d771eb410c52d65f367869b32c520c1d0c60486ef1).
Thus, the doc generator seems to depend on a previous version of `rome_js_syntax`.
Is this even possible?

**Comment by ematipico** [comment]
> Note that I added [`DOLLAR_CURLY` to the token kind -> string mapping](https://github.com/biomejs/biome/pull/70/files#diff-0065bccc1147cd025d6848d771eb410c52d65f367869b32c520c1d0c60486ef1). Thus, the doc generator seems to depend on a previous version of `rome_js_syntax`. Is this even possible?


That's a **generated** file; it's inside a **`generated/` directory**. If you want to add that change, you need to modify the following:

https://github.com/biomejs/biome/blob/b53e79c349ebc5b93017fd870172bc98b4276b4a/xtask/codegen/src/kinds_src.rs#L19-L81

File: crates/rome_js_analyze/src/analyzers/style/use_template.rs [hunk 1]
```diff
@@ -109,76 +87,140 @@ impl Rule for UseTemplate {
...
-                        [],
-                    )),
-                );
-                reduced_exprs.push(chunk_element);
+/// Returns true if `node` can be converted to a template literal.
+///
+/// This is the case, if:
+///
+/// - the operation has `+` as operator,
+/// - the operation contains a string-like literal and a non-string-like
```
Comment (ematipico) [inline crates/rome_js_analyze/src/analyzers/style/use_template.rs:112]:
```suggestion
/// - the binary expression contains the `+` operator,
/// - the binary expression contains a string-like literal and a non-string-like
```

File: crates/rome_js_analyze/src/analyzers/style/use_template.rs [hunk 2]
```diff
@@ -109,76 +87,140 @@ impl Rule for UseTemplate {
...
             }
-            AnyJsExpression::JsTemplateExpression(template) => {
-                reduced_exprs.extend(flatten_template_element_list(template.elements())?);
+            iter.skip_subtree();
+        }
+    }
+    Some(false)
+}
+
+fn template_expression_from(node: &JsBinaryExpression) -> Option<JsTemplateExpression> {
```
Comment (ematipico) [inline crates/rome_js_analyze/src/analyzers/style/use_template.rs:150]: nit: `template_expression_from_binary_expression`

File: crates/rome_js_analyze/src/analyzers/style/use_template.rs [hunk 3]
```diff
@@ -109,76 +87,140 @@ impl Rule for UseTemplate {
...
+    // While `template_elements` is empty, we keep track of last leaved node.
+    // Once we see the first string/template literal,
+    // we insert `last_leaved_node` in `template_elements` and the seen string/template literal.
+    // ANy succeeding expression is directly inserted in `template_elements`.
+    let mut template_elements = vec![];
+    let mut last_leaved_node = None;
```
Comment (ematipico) [inline crates/rome_js_analyze/src/analyzers/style/use_template.rs:156]: typo, it should be `last_left_node`

File: crates/rome_js_analyze/src/analyzers/style/use_template.rs [hunk 4]
```diff
@@ -109,76 +87,140 @@ impl Rule for UseTemplate {
...
+        }
+    }
+    Some(false)
+}
+
+fn template_expression_from(node: &JsBinaryExpression) -> Option<JsTemplateExpression> {
+    // While `template_elements` is empty, we keep track of last leaved node.
+    // Once we see the first string/template literal,
+    // we insert `last_leaved_node` in `template_elements` and the seen string/template literal.
+    // ANy succeeding expression is directly inserted in `template_elements`.
```
Comment (ematipico) [inline crates/rome_js_analyze/src/analyzers/style/use_template.rs:154]:
```suggestion
    // While `template_elements` is empty, we keep track of the last left node.
    // Once we see the first string/template literal,
    // we insert `last_left_node` in `template_elements` and the seen string/template literal.
    // Any subsequent expression is directly inserted in `template_elements`.
```

File: crates/rome_js_analyze/src/analyzers/style/use_template.rs [hunk 5]
```diff
@@ -109,76 +87,140 @@ impl Rule for UseTemplate {
...
             make::token(T!['`']),
-            make::js_template_element_list(reduced_exprs),
+            make::js_template_element_list(template_elements),
             make::token(T!['`']),
         )
         .build(),
     )
 }
 
+#[inline]
```
Comment (ematipico) [inline crates/rome_js_analyze/src/analyzers/style/use_template.rs:208]: Is it needed?

**Review by ematipico — APPROVED** [review]

**Pull request merged**